### PR TITLE
[RISCV] Move RISCVDeadRegisterDefinitions to post vector regalloc

### PR DESCRIPTION
**Pull request by lukel97**
Currently RISCVDeadRegisterDefinitions runs after vsetvli insertion, but in #70549 vsetvli insertion runs after vector regalloc and as a result we no longer convert some vsetvli a0, a0s to vsetvli x0, a0. This patch moves it to after vector regalloc, but before scalar regalloc so we still get the benefits of reducing register pressure.

There are no differences on llvm-test-suite or SPEC CPU 2017 with -O3 and -march=rv64gv, just one case in the in-tree tests where a vsetvli output VL is now marked as dead after phi elimination.

### CODE REVIEW

**Review by dtcxzyw — APPROVED** [review]
LGTM.

**Comment by topperc** [comment]
Other than vsetvli, the original patch only affected atomics and non-volatile loads. I'm not sure if those are well tested by SPEC or llvm-test-suite. Especially in high register pressure situation.

> believe the register allocator already accounts for dead defs hence why this almost an NFC.

If all the non-reserved registers are in use, the register allocation won't pick X0. It would spill to pick a non-reserved register.

**Comment by lukel97** [comment]
> If all the non-reserved registers are in use, the register allocation won't pick X0. It would spill to pick a non-reserved register.

That makes more sense, I forgot that X0 was reserved.

@BeMg I also found we have some regressions in #70549 since this no longer runs after vsetvli insertion, it seems to affect cases where two vsetvlis were coalesced e.g.:

```diff
 define <vscale x 1 x double> @test1(i64 %avl, <vscale x 1 x double> %a, <vscale x 1 x double> %b) nounwind {
 ; CHECK-LABEL: test1:
 ; CHECK:       # %bb.0: # %entry
-; CHECK-NEXT:    vsetvli zero, a0, e64, m1, ta, ma
+; CHECK-NEXT:    vsetvli a0, a0, e64, m1, ta, ma
 ; CHECK-NEXT:    vfadd.vv v8, v8, v9
 ; CHECK-NEXT:    ret
entry:
  %0 = tail call i64 @llvm.riscv.vsetvli(i64 %avl, i64 2, i64 7)
  %1 = tail call <vscale x 1 x double> @llvm.riscv.vfadd.nxv1f64.nxv1f64(
    <vscale x 1 x double> undef,
    <vscale x 1 x double> %a,
    <vscale x 1 x double> %b,
    i64 7, i64 %0)
  ret <vscale x 1 x double> %1
}
```

We could potentially run this twice, before regalloc and after regalloc. Or I think we might be able to also make coalescing smarter and have it preserve the x0 def.

**Comment by lukel97** [comment]
I've updated this to run after vector regalloc but before scalar regalloc, so we should have the best of both worlds now.

File: llvm/lib/Target/RISCV/RISCVDeadRegisterDefinitions.cpp [hunk 1]
```diff
@@ -89,8 +96,8 @@ bool RISCVDeadRegisterDefinitions::runOnMachineFunction(MachineFunction &MF) {
           LLVM_DEBUG(dbgs() << "    Ignoring, register is not a GPR.\n");
           continue;
         }
+        LIS.removeInterval(Reg);
```
Comment (BeMg) [inline llvm/lib/Target/RISCV/RISCVDeadRegisterDefinitions.cpp:99]: nit: maybe could add `if(LIS.hasInterval(Reg))` or `assert(LIS.hasInterval(Reg))` to make sure the VReg has intervals.

**Review by BeMg — APPROVED** [review]
LGTM with one nit

**Pull request merged**